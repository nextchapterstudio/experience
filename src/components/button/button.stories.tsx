import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'

import { Button } from './button'

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: 'Example/Button',
  component: Button,
} as ComponentMeta<typeof Button>

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />

export const Primary = Template.bind({})
// More on args: https://storybook.js.org/docs/react/writing-stories/args
Primary.args = {
  isPrimary: true,
  children: 'Primary Button',
}

export const Info = Template.bind({})
Info.args = {
  children: 'Info Button',
  isInfo: true,
}
export const Success = Template.bind({})
Success.args = {
  children: 'Success Button',
  isSuccess: true,
}
export const Inactive = Template.bind({})
Inactive.args = {
  children: 'Inactive Button',
  isStatic: true,
}

export const Large = Template.bind({})
Large.args = {
  isLarge: true,
  children: 'Button',
}

export const Small = Template.bind({})
Small.args = {
  isSmall: true,
  children: 'Button',
}
