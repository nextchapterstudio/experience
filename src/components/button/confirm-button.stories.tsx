import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'

import { ConfirmButton } from './'

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: 'Example/ConfirmButton',
  component: ConfirmButton,
} as ComponentMeta<typeof ConfirmButton>

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof ConfirmButton> = (args) => <ConfirmButton {...args} />

export const Primary = Template.bind({})
// More on args: https://storybook.js.org/docs/react/writing-stories/args
Primary.args = {
  children: 'Primary Button',
}
