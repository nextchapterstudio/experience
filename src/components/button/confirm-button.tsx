import React, { FC, HTMLAttributes } from 'react'
import {
  ModalCardBody,
  ModalCardFoot,
  ModalCardHead,
  Button,
  Field,
  Control,
  ModalCard,
  CardModal,
} from '../'
import { useToggle } from '../../hooks'
import { noop } from '../../utils'

export interface ConfirmButtonProps {
  onConfirm: () => void
  onCancel: () => void
  question?: string
  confirmText?: string
  declineText?: string
  isDisabled?: boolean
  title?: string
}
const ConfirmButton: FC<ConfirmButtonProps> = ({
  onConfirm = noop,
  onCancel = noop,
  question = 'Are you sure?',
  confirmText = 'OK',
  declineText = 'Cancel',
  isDisabled,
  children,
  title,
}) => {
  const [modalShown, setModalShown] = useToggle(false)

  console.log('modlal shown', modalShown)
  return (
    <>
      <Button
        isPrimary
        isLoading={modalShown}
        onClick={(e) => {
          setModalShown()
          e.preventDefault()
        }}
      >
        {children}
      </Button>
      <CardModal includeTrigger={false} isActive={modalShown} onCloseClick={onCancel}>
        <ModalCardHead title={title}></ModalCardHead>
        <ModalCardBody>{question}</ModalCardBody>
        <ModalCardFoot>
          <Field isGrouped>
            <Control>
              <Button
                isSuccess
                onClick={(e) => {
                  e.preventDefault()
                  onConfirm()
                  setModalShown()
                }}
              >
                {confirmText}
              </Button>
              <Button
                onClick={(e) => {
                  e.preventDefault()
                  onCancel()
                  setModalShown()
                }}
              >
                {declineText}
              </Button>
            </Control>
          </Field>
        </ModalCardFoot>
      </CardModal>
    </>
  )
}

ConfirmButton.defaultProps = {
  confirmText: 'Ok',
  onCancel: noop,
  onConfirm: noop,
  declineText: 'Cancel',
  question: 'Are you sure?',
  title: 'Please confirm',
}
export { ConfirmButton }
export default ConfirmButton
