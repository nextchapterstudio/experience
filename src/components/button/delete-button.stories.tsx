import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'

import { DeleteButton } from './'

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: 'Example/DeleteButton',
  component: DeleteButton,
} as ComponentMeta<typeof DeleteButton>

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof DeleteButton> = (args) => <DeleteButton {...args} />

export const Primary = Template.bind({})
// More on args: https://storybook.js.org/docs/react/writing-stories/args
Primary.args = {
  children: 'Delete Button',
}
