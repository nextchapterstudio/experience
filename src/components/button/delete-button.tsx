import React, { FC } from 'react'
import classnames from 'classnames'
import { Sizes } from '../../utils/modifiers'

export interface DeleteProps {
  className?: string
  onClick: (e: React.MouseEvent<HTMLButtonElement>) => void
  isSmall?: boolean
  isMedium?: boolean
  isLarge?: boolean
}

export const DeleteButton: FC<DeleteProps> = ({
  className,
  onClick,
  isSmall,
  isMedium,
  isLarge,
}) => {
  const classes = Sizes({ isSmall, isMedium, isLarge })
  return <button className={classnames('delete', className, classes)} onClick={onClick}></button>
}

DeleteButton.defaultProps = {
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  onClick: () => {},
}

export default DeleteButton
