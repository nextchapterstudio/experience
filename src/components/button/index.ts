export * from './button'
export * from './confirm-button'
export * from './delete-button'
