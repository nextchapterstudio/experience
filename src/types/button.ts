export interface ButtonProps {
  isPrimary?: boolean
  isSuccess?: boolean
  isInfo?: boolean
  isWarning?: boolean
  isDanger?: boolean
  isLink?: boolean
  isLight?: boolean
  isAnchor?: boolean
  isSubmit?: boolean
  isReset?: boolean
  isWhite?: boolean
  isDark?: boolean
  isBlack?: boolean
  isLoading?: boolean
  isText?: boolean
  isSmall?: boolean
  isMedium?: boolean
  isNormal?: boolean
  isLarge?: boolean
  isFullWidth?: boolean
  isInverted?: boolean
  isOutlined?: boolean
  isRounded?: boolean
  isActive?: boolean
  isHovered?: boolean
  isFocused?: boolean
  isStatic?: boolean
  isSelected?: boolean
  onClick?: (e: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement>) => void
}
