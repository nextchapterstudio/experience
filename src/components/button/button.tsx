import React, { forwardRef, ElementType, HTMLAttributes, FC, ComponentProps } from 'react'
import classnames from 'classnames'
import { Colors, Sizes, StyleOrState } from '../../utils/modifiers'
import { ButtonProps } from 'types/button'
import { ColorProps } from 'types/colors'
import htmlAttributes, { ariaAttributes, allAttributes } from '../../utils/element-attributes'

const cleanProps = (tag: string, props: any) => {
  const castedProps = props as Record<string, string>
  return htmlAttributes['*']
    .concat(htmlAttributes[tag] || [])
    .concat(ariaAttributes)
    .reduce((objs: Record<string, string>, current: string) => {
      if (castedProps[current]) {
        objs[current] = castedProps[current]
      }
      return objs
    }, {})
}

export const Button: FC<
  ButtonProps & BaseButtonProps & ColorProps & HTMLAttributes<HTMLButtonElement>
> = ({
  className,
  isPrimary,
  isSuccess,
  isInfo,
  isWarning,
  isDanger,
  isLink,
  isLight,
  isAnchor,
  isSubmit,
  isReset,
  isWhite,
  isDark,
  isBlack,
  isLoading,
  isText,
  isSmall,
  isMedium,
  isNormal,
  isLarge,
  isFullWidth,
  isInverted,
  isOutlined,
  isRounded,
  isActive,
  isHovered,
  isFocused,
  isStatic,
  isSelected,
  children,
  ...props
}) => {
  const classes = {
    ...Colors({
      isPrimary,
      isSuccess,
      isInfo,
      isWarning,
      isDanger,
      isLink,
      isLight,
      isDark,
      isBlack,
      isText,
      isWhite,
      isInverted,
      isOutlined,
      isSelected,
    }),
    ...Sizes({ isSmall, isMedium, isNormal, isLarge, isFullWidth }),
    ...StyleOrState({
      isLoading,
      isInverted,
      isOutlined,
      isRounded,
      isActive,
      isHovered,
      isFocused,
      isSelected,
    }),
  }

  if (isStatic) {
    return (
      <span className={classnames('button', className, classes)} {...cleanProps('span', props)}>
        {children}
      </span>
    )
  }
  if (isAnchor) {
    return (
      <a className={classnames('button', className, classes)} {...cleanProps('a', props)}>
        {children}
      </a>
    )
  }

  if (isSubmit) {
    return (
      <input
        className={classnames('button', className, classes)}
        {...cleanProps('input', props)}
        type="submit"
        value={children?.toString()}
      />
    )
  }
  if (isReset) {
    return (
      <input
        className={classnames('button', className, classes)}
        {...cleanProps('input', props)}
        type="reset"
        value={children?.toString()}
      />
    )
  }
  return (
    <button className={classnames('button', className, classes)} {...cleanProps('button', props)}>
      {children}
    </button>
  )
}

Button.propTypes = {}

export default Button

export interface ButtonsProps extends HTMLAttributes<HTMLButtonElement> {
  children?: React.ReactNode
  className?: string
  areSmall?: boolean
  areMedium?: boolean
  areLarge?: boolean
  hasAddons?: boolean
  isCentered?: boolean
  isRight?: boolean
}
export const Buttons: FC<ButtonsProps> = ({
  className,
  children,
  areSmall,
  areMedium,
  areLarge,
  hasAddons,
  isCentered,
  isRight,
}) => {
  const sized = {
    'are-small': areSmall,
    'are-medium': areMedium,
    'are-large': areLarge,
    'has-addons': hasAddons,
    'is-centered': isCentered,
    'is-right': isRight,
  }
  return <div className={classnames('buttons', className, sized)}>{children}</div>
}

export interface BaseButtonProps extends HTMLAttributes<HTMLButtonElement | HTMLAnchorElement> {
  as?: ElementType
  children?: React.ReactNode
  className?: string
  isSpaced?: boolean
  status?: 'inactive' | 'inProgress' | 'success' | 'failed'
  value?: string | number | Error
  href?: string
  onClick?: (e: React.MouseEvent<HTMLButtonElement | HTMLAnchorElement>) => void
}
