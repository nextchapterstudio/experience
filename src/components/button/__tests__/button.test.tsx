import React from 'react'
import { render, cleanup, fireEvent, screen } from '@testing-library/react'
import { Button, Buttons } from '../button'

describe('The Button Component', () => {
  afterEach(cleanup)
  it('should render with the class basic', () => {
    const { container } = render(<Button className="basic">test</Button>)

    expect(container).toMatchSnapshot()
    expect(container.firstChild).toMatchSnapshot()
    expect(container.firstChild).toHaveClass('basic')
  })

  it('should render a span when isStatic', () => {
    const { container } = render(<Button isStatic>test</Button>)

    expect(container).toMatchSnapshot()
    expect(container.firstChild).toMatchSnapshot()
  })
  it('should render a button with an aria attribute', () => {
    const { container } = render(<Button aria-hidden>test</Button>)

    expect(container).toMatchSnapshot()
    expect(container.firstChild).toHaveAttribute('aria-hidden')
  })
  it('should render as an anchor tag', () => {
    const { container } = render(<Button isAnchor>test</Button>)
    expect(container).toMatchSnapshot()
    expect(container).toContainHTML('<a')
  })
  it('should render as an input tag when isSubmit', () => {
    const { container } = render(<Button isSubmit>test</Button>)
    expect(container).toMatchSnapshot()
    expect(container).toContainHTML('<a')
  })
  it('should support an onclick function', async () => {
    const mockFn = jest.fn()
    const { container } = render(<Button onClick={mockFn}>The Button</Button>)

    const b = await screen.findByText('The Button')
    await fireEvent.click(b)
    expect(container).toMatchSnapshot()
    expect(mockFn).toBeCalled()
  })
  it('should render the Buttons with colored classes applied', () => {
    const { container } = render(
      <Buttons>
        <Button isPrimary>Primary</Button>
        <Button isInfo>Info</Button>
        <Button isDanger>Danger</Button>
        <Button isLink>Link</Button>
        <Button isWarning>Warning</Button>
        <Button isSuccess>Success</Button>
      </Buttons>
    )
    expect(container.firstChild).toMatchSnapshot()
    expect(container.firstChild).toHaveClass('buttons')
  })
  it('should render the Buttons as correct elements', () => {
    const { container } = render(
      <Buttons>
        <Button>Button</Button>
        <Button isSubmit>Submit</Button>
        <Button isAnchor href="#">
          Anchor
        </Button>
        <Button isReset>Reset</Button>
      </Buttons>
    )
    expect(container.firstChild).toMatchSnapshot()
  })
})
