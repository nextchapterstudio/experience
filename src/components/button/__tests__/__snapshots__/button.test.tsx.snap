// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Button Component should render a button with an aria attribute 1`] = `
<div>
  <button
    aria-hidden="true"
    class="button"
  >
    test
  </button>
</div>
`;

exports[`The Button Component should render a span when isStatic 1`] = `
<div>
  <span
    class="button"
  >
    test
  </span>
</div>
`;

exports[`The Button Component should render a span when isStatic 2`] = `
<span
  class="button"
>
  test
</span>
`;

exports[`The Button Component should render as an anchor tag 1`] = `
<div>
  <a
    class="button"
  >
    test
  </a>
</div>
`;

exports[`The Button Component should render as an input tag when isSubmit 1`] = `
<div>
  <input
    class="button"
    type="submit"
    value="test"
  />
</div>
`;

exports[`The Button Component should render the Buttons as correct elements 1`] = `
<div
  class="buttons"
>
  <button
    class="button"
  >
    Button
  </button>
  <input
    class="button"
    type="submit"
    value="Submit"
  />
  <a
    class="button"
    href="#"
  >
    Anchor
  </a>
  <input
    class="button"
    type="reset"
    value="Reset"
  />
</div>
`;

exports[`The Button Component should render the Buttons with colored classes applied 1`] = `
<div
  class="buttons"
>
  <button
    class="button is-primary"
  >
    Primary
  </button>
  <button
    class="button is-info"
  >
    Info
  </button>
  <button
    class="button is-danger"
  >
    Danger
  </button>
  <button
    class="button is-link"
  >
    Link
  </button>
  <button
    class="button is-warning"
  >
    Warning
  </button>
  <button
    class="button is-success"
  >
    Success
  </button>
</div>
`;

exports[`The Button Component should render with the class basic 1`] = `
<div>
  <button
    class="button basic"
  >
    test
  </button>
</div>
`;

exports[`The Button Component should render with the class basic 2`] = `
<button
  class="button basic"
>
  test
</button>
`;

exports[`The Button Component should support an onclick function 1`] = `
<div>
  <button
    class="button"
  >
    The Button
  </button>
</div>
`;
